src/aosm/azext_aosm/generate_nsd/nsd_generator.py

```diff
@@ -0,0 +1,212 @@
+    """
+    NSD Generator.
+
+    This takes a config file and a set of NFDV deploy_parameters and outputs:
+    - A bicep file for the NSDV
+    - Parameters files that are used by the NSDV bicep file, these are the
+      schemas and the mapping profiles of those schemas parameters
+    - A bicep file for the Artifact manifest
+    - A bicep and JSON file defining the Network Function that will
+      be deployed by the NSDV
+    """
+
+    def __init__(self, config: NSConfiguration):
+        self.config = config
+        self.nsd_bicep_template_name = NSD_DEFINITION_BICEP_SOURCE_TEMPLATE
+        self.nf_bicep_template_name = NF_TEMPLATE_BICEP_FILE
+        self.nsd_bicep_output_name = NSD_DEFINITION_BICEP_FILE
+
+        self.build_folder_name = self.config.build_output_folder_name
+
+    def generate_nsd(self, deploy_parameters) -> None:
+        """Generate a NSD templates which includes an Artifact Manifest, NFDV and NF templates."""
+        logger.info(f"Generate NSD bicep template")
+
+        self.deploy_parameters = deploy_parameters
+
+        self._create_nsd_folder()
+        self.create_parameter_files()
+        self.write_nsd_manifest()
+        self.write_nf_bicep()
+        self.write_nsd_bicep()
+
+        print(f"Generated NSD bicep templates created in {self.build_folder_name}")
+        print(
+            "Please review these templates. When you are happy with them run "
+            "`az aosm nsd publish` with the same arguments."
+        )
+
+    def _create_nsd_folder(self) -> None:
+        """
+        Create the folder for the NSD bicep files.
+
+        :raises RuntimeError: If the user aborts.
+        """
+        if os.path.exists(self.build_folder_name):
+            carry_on = input(
+                f"The folder {self.build_folder_name} already exists - delete it and continue? (y/n)"
+            )
+            if carry_on != "y":
+                raise RuntimeError("User aborted!")
+
+            shutil.rmtree(self.build_folder_name)
+
+        logger.info("Create NFD bicep %s", self.build_folder_name)
+        os.mkdir(self.build_folder_name)
+
+    def create_parameter_files(self) -> None:
+        """Create the Schema and configMappings json files."""
+        schemas_folder_path = os.path.join(self.build_folder_name, SCHEMAS)
+        os.mkdir(schemas_folder_path)
+        self.write_schema(schemas_folder_path)
+
+        mappings_folder_path = os.path.join(self.build_folder_name, CONFIG_MAPPINGS)
+        os.mkdir(mappings_folder_path)
+        self.write_config_mappings(mappings_folder_path)
+
+    def write_schema(self, folder_path: str) -> None:
+        """
+        Write out the NSD Config Group Schema JSON file.
+
+        :param folder_path: The folder to put this file in.
+        """
+        logger.debug(f"Create {self.config.cg_schema_name}.json")
+
+        schema_path = os.path.join(folder_path, f"{self.config.cg_schema_name}.json")
+
+        with open(schema_path, "w") as _file:
```

Comment:
this should be formatted better, on a json dump you can do indent=4 as a param, im not sure how you would do it in this case